File: ruby/16_3sum_closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}

# 给定数组和target，求数组中三元组和最接近target的
# 思路和3Sum第二种方法类似，首先排序，求数组中前三个的和和target之前的distance
# 然后遍历数组，依次比较各个三元组和target之前的dis，最后取得dis最小的，返回target+dis即可


def three_sum_closest(nums, target)
    nums = nums.sort

    distance = nums[0..2].reduce(:+) - target

    0.upto(nums.length-3) do |ai|
    	bi, ci = ai + 1, nums.length-1

    	while bi < ci
    		sum = nums[ai] + nums[bi] + nums[ci]

    		dis = sum - target
    		return sum if dis == 0

    		distance = dis if dis.abs < distance.abs

    		bi += 1 if dis < 0
    		ci -= 1 if dis > 0
    	end
    end

    target + distance
end


nums = [1, 3, 5, 7, 9]

File: cpp/16_3sum_closest.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 给定数组和target，求三个一组和最接近target的和
// 将数组排序，从头开始遍历，每次left=i+1, right=size-1，每次根据sum求diff，返回diff最小的sum即可
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
		int res = nums[0] + nums[1] + nums[2];
		int diff = abs(res - target);
	
		sort(nums.begin(), nums.end());
		for (int i = 0; i < nums.size(); ++i) {
			int l = i+1, r = nums.size()-1;
			
			while (l < r) {
				int sum = nums[i] + nums[l] + nums[r];
				if (sum == target) {
					return target;
				} else if (sum > target) {
					r--;	
				} else {
					l++;
				}
				int tmp = abs(sum-target);
				if (tmp < diff) {
					diff = tmp;
					res = sum;
				}
			}
		}

		return res;
    }
};

void unit_test() {
	Solution s;

	vector<int> nums = {-1, 2, 1, -4};

	assert(s.threeSumClosest(nums, 1) == 2);
}

int main() {
	unit_test();
	return 0;
}
